Dedupe process in grpc reporter

pavolloffay opened the pull request:
Related to https://github.com/jaegertracing/jaeger/issues/773#issuecomment-436602664 and https://github.com/jaegertracing/jaeger/pull/1165#discussion_r232053284

Signed-off-by: Pavol Loffay <ploffay@redhat.com>

yurishkuro commented on cmd/agent/app/reporter/grpc/reporter.go:
```diff
@@ -55,14 +53,14 @@ func (r *Reporter) EmitZipkinBatch(zSpans []*zipkincore.Span) error {
+	return r.send(trace.Spans, process)
+}
+
+func (r *Reporter) send(spans []*model.Span, process model.Process) error {
```
Maybe we should pass process=nil for zipkin path? I don't recall how collector handles batches, it should not apply batch-level Process to each span if spans already have a Process attached. This would be more compatible with, say, Istio use case where one client can emit spans for different processes.

Once we do that, we can kill this zipkin code path in the processor API and pull it higher in the chain, i.e. since you're already calling toDomain in the grpc package (but not in tchannel package), if we confirm that we don't cause a data loss we can call toDomain before invoking the processor.

NB: we may still need to keep the zipkin tchannel endpoint in the collector because some very old clients may be using direct tchannel connections to the collector, although I find it hard to believe since the agent was implemented long before we open sourced Jaeger backend (like 1yr before), so nobody outside of Uber should be using those old clients. If so, we can remove that endpoint and re-keep it in our internal build only.

pavolloffay replied:
>  I don't recall how collector handles batches

```go
// PostSpans implements gRPC CollectorService.
func (g *GRPCHandler) PostSpans(ctx context.Context, r *api_v2.PostSpansRequest) (*api_v2.PostSpansResponse, error) {
	for _, span := range r.GetBatch().Spans {
		if span.GetProcess() == nil {
			span.Process = &r.Batch.Process
		}
	}
	_, err := g.spanProcessor.ProcessSpans(r.GetBatch().Spans, JaegerFormatType)
	if err != nil {
		g.logger.Error("cannot process spans", zap.Error(err))
		return nil, err
	}
	return &api_v2.PostSpansResponse{}, nil
}
```

It sets the process only when it is missing.

Another issue is that this zipkin path skips zipkin sanitizers which are used in collector handler.

pavolloffay replied:
> Once we do that, we can kill this zipkin code path in the processor API and pull it higher in the chain, i.e. since you're already calling toDomain in the grpc package (but not in tchannel package), if we confirm that we don't cause a data loss we can call toDomain before invoking the processor.

I don't get you here. Once we do what pass the nill process? I have just added sanitizers directly to the agent.


> so nobody outside of Uber should be using those old clients. If so, we can remove that endpoint and re-keep it in our internal build only.
tracked in  https://github.com/jaegertracing/jaeger/issues/1186 for

yurishkuro replied:
Right now we have this flow:
```
udp server jaeger.thrift -> reporter::SubmitBatch
udp server zipking.thrift -> reporter::SubmitZipkinSpans

tchannel reporter::SubmitBatch -> pass through to collector
tchannel reporter::SubmitZipkinSpans -> pass through to collector

grpc reporter::SubmitBatch -> convert to domain and send to collector
grpc reporter::SubmitZipkinSpans -> convert to jaeger.thrift, then to domain and send to collector
```
What I am proposing (could be done in a different PR), is to eliminate `reporter::SubmitZipkinSpans` completely and have this:
```
udp server jaeger.thrift -> reporter::SubmitBatch
    tchannel reporter::SubmitBatch -> pass through to collector
    grpc reporter::SubmitBatch -> convert to domain and send to collector
udp server zipking.thrift -> convert to jaeger.thrift -> reporter::SubmitBatch
```
This way the Reporter only has a single method for Jaeger spans (in Thrift, for now).

Once we remove tchannel completely, we can push thrift->domain conversion into UDP servers as well, and Reporter interface effectively becomes == Collector service from proto IDL.

pavolloffay replied:
I see, it sounds reasonable to me. I would prefer a separate PR to do it as it might be bigger than I think right now.

pavolloffay replied:
I have created to track it separately https://github.com/jaegertracing/jaeger/issues/1193

yurishkuro commented on cmd/agent/app/reporter/grpc/reporter.go:
```diff
@@ -55,14 +53,14 @@ func (r *Reporter) EmitZipkinBatch(zSpans []*zipkincore.Span) error {
 	var process model.Process
-	if len(spans) > 0 {
-		process = *spans[0].Process
+	if len(trace.GetSpans()) > 0 {
```
nit: could make it nicer in Go; `if spans := trace.GetSpans(), len(spans) > 0 { use spans }`

yurishkuro commented on cmd/agent/app/reporter/grpc/reporter.go:
```diff
@@ -32,38 +33,46 @@ import (
+		zipkin2.NewSpanDurationSanitizer(),
+		zipkin2.NewSpanStartTimeSanitizer(),
+		zipkin2.NewParentIDSanitizer(),
+		zipkin2.NewErrorTagSanitizer(),
```
damn, just noticed that we internally didn't use all 4:
```go
func (config *Config) NewZipkinSanitizer(metrics metrics.Factory, logger *zap.Logger) zs.Sanitizer {
        sanitizers := zs.NewChainedSanitizer(
                zs.NewSpanDurationSanitizer(),
                zs.NewParentIDSanitizer(),
        )
        return &sanitizers
}
```
Can we create `zipkin2.StandardSanitizers` array instead, and then here only do this?
```go
zSanitizer := zipkin2.NewChainedSanitizer(zipkin2.StandardSanitizers...)`
```

yurishkuro commented on cmd/agent/app/reporter/grpc/reporter.go:
```diff
@@ -32,38 +33,46 @@ import (
+func (r *Reporter) send(spans []*model.Span, process *model.Process) error {
+	batch := model.Batch{Spans: spans}
+	if process != nil {
+		batch.Process = *process
```
oh-oh, I think I made a mistake in the model:
```
type Batch struct {
        Spans   []*Span `protobuf:"bytes,1,rep,name=spans" json:"spans,omitempty"`
        Process Process `protobuf:"bytes,2,opt,name=process" json:"process"`
}
```
Considering the possibility of Zipkin spans, or even Istio spans with mixed services, the Process field should be optional: `*Process`. Should we make this change?  It's not an IDL change, just the Gogo annotations affecting Go struct.

pavolloffay replied:
changed to 
```proto
message Batch {
    repeated Span spans = 1;
    Process process = 2 [
      (gogoproto.nullable) = true
    ];
}
```

pavolloffay commented:
I think I have addressed all comments. I have created https://github.com/jaegertracing/jaeger/issues/1193 to remove `EmitBatch`. If there are no other comments I think it could go in and also the metrics PR https://github.com/jaegertracing/jaeger/pull/1180. The #1193 will handle removal of duplicated metrics e.g.

```
jaeger_agent_tchannel_reporter_batches_submitted{format="jaeger"} 0
jaeger_agent_tchannel_reporter_batches_submitted{format="zipkin"} 0
```

yurishkuro commented on cmd/collector/app/sanitizer/zipkin/span_sanitizer.go:
```diff
@@ -28,6 +28,8 @@ const (
 var (
 	defaultDuration = int64(1)
+	// StandardSanitizers is a list of standard zipkin sanitizers.
+	StandardSanitizers = []Sanitizer{NewSpanStartTimeSanitizer(), NewSpanDurationSanitizer(), NewParentIDSanitizer(), NewErrorTagSanitizer()}
```
nit: one per line

yurishkuro commented on model/proto/model.proto:
```diff
@@ -155,6 +155,6 @@ message Trace {
     repeated Span spans = 1;
     Process process = 2 [
-      (gogoproto.nullable) = false
+      (gogoproto.nullable) = true
```
nit: I would remove the option altogether (true is the default afaik)
